### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by svetlanagorchanina**
**EVALUATION CRITERIA**

- Task 4.1 is implemented
- TASK 4.2 is implemented lambda links are provided and returns data
- TASK 4.3 is implemented lambda links are provided and products is stored in DB
- Frontend application is integrated with product service

**Additional**

- POST/products lambda functions returns error 400 status code if product data is invalid
- All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
- All lambdas do console.log for each incoming requests and their arguments
-  Transaction based creation of product

---

FE MR: https://github.com/svetlanagorchanina/nodejs-aws-fe/pull/3
BE MR: https://github.com/svetlanagorchanina/nodejs-aws-be/pull/2
FE App link: https://d12r4l4twrn36b.cloudfront.net/
Product Servece API:
- GET product list: https://fgkaarac1j.execute-api.eu-west-1.amazonaws.com/dev/products
- GET product: https://fgkaarac1j.execute-api.eu-west-1.amazonaws.com/dev/products/34db16b8-2638-4c8c-89db-2fde30f66459
- POST product: https://fgkaarac1j.execute-api.eu-west-1.amazonaws.com/dev/products

---

Notes:
- run test via`npm run test`
- for swagger documentation use `swagger.json`

### CODE REVIEW

**Comment by OHalahan** [comment]
Lambda for creating product is not integrated with Frontend

image

Just change **axios.put** to **axios.post** in frontend app

Same goes for getting list of products https://d12r4l4twrn36b.cloudfront.net/admin/products
Just correct paths to API

File: product-service/handlers/createProduct/index.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        statusCode: 400,
+        body: JSON.stringify({ message: "Product is not valid" }),
+      });
+    }
+
+    try {
+      client.query("BEGIN");
+      const {
+        rows: [{ id }],
+      } = await client.query(
```
Comment (OHalahan) [inline product-service/handlers/createProduct/index.ts]: from security perspective it is more safer to pass parameters as `client.query('query string', [param1, param2])`.
Injecting values directly into query string opens SQL injection vulnerabilities.

This does not affect final grade by any means, but Transaction based product creation looks almost perfect here except this part

**Comment by svetlanagorchanina** [comment]
> Lambda for creating product is not integrated with Frontend
> 
> image
> 
> Just change **axios.put** to **axios.post** in frontend app
> 
> Same goes for getting list of products https://d12r4l4twrn36b.cloudfront.net/admin/products
> Just correct paths to API

@OHalahan  Thanks for noticing! All related changes were implemented in https://github.com/svetlanagorchanina/nodejs-aws-fe/pull/3.

File: product-service/handlers/createProduct/index.ts (hunk 1 above)
Comment (svetlanagorchanina) [reply product-service/handlers/createProduct/index.ts]: @OHalahan Great point, PR was updated

**Review by OHalahan — APPROVED** [review]

**Review by loginov-rocks — APPROVED** [review]
Nice work!

Additional idea: for the wrong UUID it might make sense to return 404 Not Found rather than 500: https://fgkaarac1j.execute-api.eu-west-1.amazonaws.com/dev/products/wrong-id

**Pull request merged**